refactor: modernize repo

Yberion opened the pull request:
Hello,

I updated the repo with up to date Angular practices, config and struct.
I separated the work in many commits when possible.
I guess it will be a bit opinionated, feel free to tell me if anything feels "wrong" for you and I'll revert the specific commit.
I will also try to explain each commit when it's needed (not all commit has description, title might be enough).

**Questions:**

- Is it a static site or do you need SSR?
  - There's only a single route, so I expect SSG to be enough, but I don't know how works Netlify and what's your using on it
  - If you need SSR, I'll migrate to the new server API
  - **Edit:** look like SSR not needed :p
- Can you test that it works properly?
  - I was not able to test it on november lul

`refactor: lock package versions to avoid breaking changes`

I encoutered breaking changes even on `patch` changes a lot of time, to avoid that let's lock the versions, this will require explicite update.

`refactor: remove eslint`

I removed eslint configs here because the config was a bit old, I will add it back from scratch later with up to date config.

`build: update packages`

I did not update `typescript` to latest as Angular 19 currently support version `5.6`.

`refactor(package.json): remove npx usage in scripts`

This is not needed and force to use `npm`.

`refactor: remove husky as it is currently not used`

I will reinstall it from scratch later as there was some changes since version `9`.

`refactor: remove karma/jasmine`

Tests were removed with the Angular 19 migration (they were empty).

`refactor: remove @angular/forms`

This was not used

`refactor: migrate to pnpm`

It is faster and stricter.

Idk what will happen on Netlify, maybe you'll have to update things there.
I've updated the readme for installation instructions

`refactor: update tsconfig`

Updated to match fresh install + your custom settings.

`build: install serve and add a script to execute it on the bundle`

This will allow us to test the built app as if it was deployed as a static site, this will help me to check if I did break anything.

`refactor: update angular.json & co`

We're now using the `public` folder.
The `assets` folder is still being used for the background as the image will be preprocessed by the css preprocessor.
We're now using the new `outputMode` with the value `static`. (I asked a question regarding if you require SSR or not, there's only 1 page, so SSR might be useless)
`server.ts` is currently not used as we're using `static` as `outputMode`.
`server.ts` was moved inside the `src` folder, same as on a fresh app.

`refactor: update the packt logo`

The logo is now in the `public` folder and is directly comming from the official website.

`refactor: update .prettierrc.json & format repo`

I renamed the file to `.prettierrc.json` as it's a json file.
I added `printWidth` to `140`.
I removed some value that were already set by default, it match a bit more nx's default config.

`perf: reorder head in index.html to more closely match Capo.js order`

I couldn't reorder the appended css elements in the `head` as it's appended by Angular during the compilation.

`feat: update app config to use eventCoalescing & withEventReplay`

I can remove event replay if needed.

`refactor: add eslint & lint src folder & fix problems`

I use the nx default config as it's good enough.
I also added some extra rules (some are using typechecking), some might be opinionated.

~~There's a warning when installing package related to `angular-eslint`, it's currently [work in progress](https://github.com/angular-eslint/angular-eslint/pull/2109) for Angular 19 support.~~

`refactor: add husky` & `refactor: add nano-staged with lint and format`

I'm using `nano-staged` instead of `lint-staged` in effort to [e18e](https://e18e.dev/) ([blog post](https://e18e.dev/blog/october-contributions-showcase.html#nano-staged)) ([e18e doc for lint-staged](https://github.com/es-tooling/module-replacements/blob/main/docs/modules/lint-staged.md)).

It does not check for the format, it simply does it.

`refactor: use readonly on properties that are initialized by Angular`

From the new [style guide](https://gist.github.com/jelbourn/0158b02cfb426e69c172db4ec92e3c0c#use-readonly-on-properties-that-are-initialized-by-angular).

`refactor: use protected on class members that are only used by a component's template`

From the new [style guide](https://gist.github.com/jelbourn/0158b02cfb426e69c172db4ec92e3c0c#use-protected-on-class-members-that-are-only-used-by-a-components-template).

`refactor: zoneless`

Remove the use of `runOutsideAngular` as well.

`perf: optimize image weight`

I moved copy of unoptimized images inside the folder `unoptimized_images` just in case.
I used `Tinypng` to reduce up to 80% the weight of the replaced images.

danielglejzner commented:
Hey! Looks huge 😃
Thank you for the contribution - I'm going to do my best to get it reviewed today + take a look myself

Yberion commented on src/app/feature/calendar/dialogs/card-details-dialog/card-details-dialog.component.html:
```diff
@@ -62,7 +62,7 @@ <h1>Remember to click on everyone 🙃🎅🎄</h1>
   <div class="xmas-dialog-content xmas-dialog-select-expert">
     @for (content of data.contents; track $index) {
-      <div class="xmas-dialog-card-wrapper" (click)="selectedContent.set(content)">
+      <div role="presentation" class="xmas-dialog-card-wrapper" (click)="selectedContent.set(content)">
```
Not sure about this one, couldn't properly test to pick the best solution

Yberion commented on src/app/feature/calendar/core/snowflakes/snowflakes.component.ts:
```diff
@@ -32,18 +31,16 @@ export class SnowflakesComponent {
   }
 
   private letItSnow(): void {
-    this.zone.runOutsideAngular(() => {
```
Still a bit unclear for me if this is needed to be removed as currently zoneless is kind of "hybrid" right now.
Angular Material is zoneless compatible but still contains a lot of `runOutsideAngular`.

Maybe I should add it back?

Any input on this @JeanMeche?

JeanMeche replied:
This app still relies on Zone.js, but it doesn't a call on a patch API. This should be OK.

Yberion replied:
Well I removed zone.js in [this commit](https://github.com/danielglejzner/angular-christmas-calendar/pull/86/commits/0dcb4856db3b473ed6c86438b1ee94382d8e9070) ^^

JeanMeche replied:
Oh perfect!
 
Also libraries will still rely on `runOutsideAngular` when they want to support both zoneless & zonefull apps. 
For app, `runOutsideAngular` is a noop when we use the zoneless provider.

Yberion replied:
Ohhhhhhh okay, I get it I think, thanks!

Yberion commented on src/app/feature/calendar/core/snowflakes/snowflakes.component.ts:
```diff
@@ -1,60 +1,49 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class SnowflakesComponent {
   @ViewChild('snowArea') snowArea!: ElementRef<HTMLSpanElement>;
```
Oh, missed this one :D

Yberion commented on src/app/feature/calendar/cards/calendar-card-factory.directive.ts:
```diff
@@ -57,12 +58,12 @@ export const christmasCardsFactoryMap: Record<ChristmasCardType, Type<ChristmasC
   standalone: true,
```
Oh indded, I though it was already done by the Angular 19 migration on the previous PR heh

Yberion commented on src/server.ts:
```diff
@@ -1,24 +1,22 @@
+import bootstrap from './main.server';
 
-const commonEngine = new CommonEngine()
+const commonEngine = new CommonEngine();
```
Well, SSR seems to be useless here (there's only 1 route that is already being SSG), it was part of my question on the main message `If you need SSR, I'll migrate to the new server API` :p

I'll probably do it anyway, it will force to use the new `ServerRoute` to setup the SSG as `prerender` is not being used with `outputMode` from what I understood

Yberion commented:
I've changed the `outputMode` to `server`.
I re-did the `server.ts` using Fastify instead of Express (strange behavior when defining route path with Express 5).

I added a fallback route that will redirect to the home page to avoid some error such error:

![image](https://github.com/user-attachments/assets/e8404853-3c2c-47fa-bd6c-4745baa37895)

Idk how it will act on Netlify's "edge function".

Yberion commented:
Ok, it seems that it require a specific request handler for Netflity (edge in general I guess).
I've tried to support both Netlify and Node.

It seems to still crash, hm, I will switch back to Express 4 in case there's a problem with Fastify.

Kordrad commented on src/app/feature/calendar/calendar-grid.component.ts:
```diff
@@ -1,27 +1,27 @@
 })
 export class CalendarGridComponent {
   private readonly calendarService = inject(CalendarService);
-  cards = this.calendarService.cards;
+  protected cards = this.calendarService.cards;
```
Maybe just like this:
```ts
  protected readonly cards = inject(CalendarService).cards;
```

You can avoid private members above public in this way

Yberion replied:
@danielglejzner would disagree with you https://x.com/DanielGlejzner/status/1861420332352991507, that's why I did not change this.

Kordrad replied:
Thanks for sharing! I'll look at it later, if I learn something new I will reopen the thread

danielglejzner replied:
Many different opinions on this one 😎 - I prefer to keep service private readonly

Kordrad commented on src/app/feature/calendar/cards/calendar-card-factory.directive.ts:
```diff
@@ -54,15 +55,14 @@ export const christmasCardsFactoryMap: Record<ChristmasCardType, Type<ChristmasC
 })
 export class CalendarCardFactoryDirective implements OnInit {
-  @Input({required: true}) type!: ChristmasCardType;
+  readonly type: InputSignal<ChristmasCardType> = input.required<ChristmasCardType>();
```
I think you don't need type twice

```suggestion
  readonly type: InputSignal<ChristmasCardType> = input.required();
```

Yberion replied:
That my personal preference, I can remove the generic type or let infer the type if you really want this change?

Yberion replied:
Sure I will change this ^^

Kordrad commented on src/app/feature/calendar/cards/candycane/candycane.component.ts:
```diff
@@ -1,13 +1,12 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class CandycaneComponent implements ChristmasCard {
   type = ChristmasCardType.Candycane;
```
you can use `readonly`

Yberion replied:
This property seems not to be used, is it safe to remove it (on all components) or did I miss something? @danielglejzner

danielglejzner replied:
Need to double check - I have no memory on that. Feel free to test it locally with .json file. I might do it today.

Yberion replied:
Yeah I've tested in local and it still works, I will push a commit that remove them.

![image](https://github.com/user-attachments/assets/b4341b0f-cdd3-44b6-a9e3-9af9e326d495)

Kordrad commented on src/app/feature/calendar/cards/christmas-tree/christmas-tree.component.html:
```diff
@@ -1149,42 +1146,15 @@
+        <div class="star" style="transform: translateZ(225px) translateY(-5.5px) rotateX(-90deg)"></div>
+        <div class="star" style="transform: translateZ(225px) translateY(-5.75px) rotateX(-90deg)"></div>
+        <div class="star" style="transform: translateZ(225px) translateY(-6px) rotateX(-90deg)"></div>
+        <div class="star" style="transform: translateZ(225px) translateY(-6.25px) rotateX(-90deg)"></div>
```
`@for`? 

I see these stars have incremental `translateY` property. You can do it way better.

Yberion replied:
Same would apply for the entire component I think, such refactor would be better in another PR I guess as this one is mostly about migrations ^^

Kordrad commented on src/index.html:
```diff
@@ -92,53 +81,96 @@
+              aria-label="Daniel Glejzner account on X"
+              target="_blank"
+              rel="nofollow"
+              style="color: #444; text-decoration: none"
```
move these styles to css file

Yberion replied:
Same as abose, this would fit more in a refactor in another PR.

Also, this CSS inlining should be taken in consideration, if it's not inlined it could be delayed by the loading of the style file and idk if it could impact SEO concerns, if the footer is directly in index.html, I think there's a reason (I myself don't know why haha)

danielglejzner replied:
I don't think there is any sophisticated thought backing this up it's just me being lazy and wanted to deliver this asap today 😂

Kordrad replied:
> Same as abose, this would fit more in a refactor in another PR.
> 
> Also, this CSS inlining should be taken in consideration, if it's not inlined it could be delayed by the loading of the style file and idk if it could impact SEO concerns, if the footer is directly in index.html, I think there's a reason (I myself don't know why haha)

Of course, inline styles have the best performance, but I think that doesn't affect too much 😄

Yberion commented on package.json:
```diff
@@ -6,51 +6,48 @@
+    "eslint": "9.16.0",
+    "eslint-config-prettier": "9.1.0",
+    "husky": "9.1.7",
+    "nano-staged": "0.8.0",
```
There was some commits recently that did not require a new version.
Unless there's security concern, if such package works and does not require new version, I'm fine with using it, it's also recommanded by e18e.

Kordrad replied:
Thanks!

Yberion commented:
Hello,

I've rebased.

Kordrad reviewed (approved):
Great job!
